### PR TITLE
refactor(api): split api/workspaces.py 1,407 → 142 LOC into services/ (closes #25)

### LINKED ISSUE #25: Monolith extraction: split api/workspaces.py first module
## Problem

`api/workspaces.py` is 1,408 LOC and co-locates database access, multi-stage dict transformation, workspace resolution, project-assignment heuristics, CLI session handling, and HTTP response formatting in a single file. The `get_workspace_tabs` route handler alone spans lines 956-1407 (~450 lines). `_determine_project_for_conversation` (lines 269-390) contains 120+ lines of cascading fallback heuristics. Helpers are duplicated across files: `_read_json_file` is defined identically in `api/workspaces.py:66` and `api/composers.py:20`. Any change to workspace logic risks regressing conversation rendering.

## Severity

High · 8 points

## Acceptance criteria

- [ ] DB helpers (`_open_global_db`, `_collect_workspace_entries`, `_build_composer_id_to_workspace_id`, `_collect_invalid_workspace_ids`) extracted to `services/workspace_db.py`
- [ ] Resolution logic (`_determine_project_for_conversation`, `_infer_invalid_workspace_aliases`, `_create_project_name_to_workspace_id_map`, `_create_workspace_path_to_id_map`, `_get_workspace_display_name`, `_infer_workspace_name_from_context`) extracted to `services/workspace_resolver.py`
- [ ] CLI workspace tab logic (`_get_cli_workspace_tabs`) extracted to `services/cli_tabs.py`
- [ ] `_read_json_file` deduplicated to a single shared location
- [ ] `api/workspaces.py` route handlers contain only request parsing, service calls, and response formatting — no direct SQLite access or dict transformation logic (target <300 LOC, down from 1,408)
- [ ] All existing tests pass without modification (behavior-preserving refactor; JSON shapes unchanged)
- [ ] CI green
- [ ] PR approved by at least 1 reviewer

### PR DESCRIPTION
**Pull request by timon0305**
Closes #25.

## What

Splits `api/workspaces.py` from **1,407 → 142 LOC** into focused service modules. Three thin route handlers remain in `api/workspaces.py`; everything else moves to `services/` and `utils/workspace_descriptor.py`. Behaviour-preserving — every existing test passes without modification.

## Why

Per the eval (Section 5.2 — *Monolith Handler*), `api/workspaces.py` was a 1,400+ LOC monolith co-locating SQLite reads, dict transformation, workspace resolution, project-assignment heuristics, CLI session handling, and HTTP response formatting in one file. `get_workspace_tabs` alone was ~450 lines. `_read_json_file` was duplicated identically in `api/composers.py:20`. The file was resistant to unit testing and a single change to workspace logic risked regressing conversation rendering.

## How

**Extraction layout** (in dependency order — pure helpers first, route bodies last):

| New module | Functions moved from `api/workspaces.py` |
|---|---|
| `utils/workspace_descriptor.py` | `_basename_from_pathish`, `_resolve_workspace_descriptor`, `_uri_or_path_to_fs_path`, `_read_json_file` (dedup from `api/composers.py:20`) |
| `services/workspace_db.py` | `_collect_workspace_entries`, `_collect_invalid_workspace_ids`, `_build_composer_id_to_workspace_id`, `_open_global_db` |
| `services/workspace_resolver.py` | `_get_workspace_display_name`, `_infer_workspace_name_from_context`, `_get_project_from_file_path`, `_create_project_name_to_workspace_id_map`, `_create_workspace_path_to_id_map`, `_determine_project_for_conversation`, `_infer_invalid_workspace_aliases` |
| `services/cli_tabs.py` | `_get_cli_workspace_tabs` |
| `services/workspace_listing.py` | body of `GET /api/workspaces` (`list_workspace_projects(...)`) |
| `services/workspace_tabs.py` | body of `GET /api/workspaces/<id>/tabs` (`assemble_workspace_tabs(...)`) |

**`api/workspaces.py` is now 142 LOC**: three route handlers that resolve the request, hand off to a service, and JSONify the result.

**Back-compat shim**: three existing tests import `_infer_invalid_workspace_aliases`, `_determine_project_for_conversation`, and `_infer_workspace_name_from_context` directly from `api.workspaces`. Those names are re-exported (with `# noqa: F401`) so the tests pass unchanged. Production callers should import from `services.workspace_resolver`.

## Verification

| Check | Result |
|---|---|
| `python3 -m unittest discover tests -v` | **186 pass** (178 existing without modification + 8 new) |
| App boots, all routes register | ✅ |
| Live smoke on my Cursor data | `/api/workspaces` → 2 workspaces, `/api/workspaces/global/tabs` → 5 tabs, `/api/search?q=test` → 7 results — all 200, zero exceptions |

## Side effects worth flagging

- **`api/composers.py` `_read_json_file` was simpler than `api/workspaces.py`'s** — it didn't handle Cursor's `{"workspace": "..."}` indirection. The consolidated version (in `utils/workspace_descriptor.py`) does. Workspaces that store an indirected `workspace.json` now resolve correctly in `/api/composers` as well — a latent bug fixed via dedup.

## CodeRabbit review pass

Eight findings addressed in this PR's follow-up commits:
1. `cli_tabs.py` — wrap `messages_to_bubbles()` in try/except so one malformed session doesn't 500 the whole endpoint
2. `workspace_db.py` — use `Path(...).resolve().as_uri()` for SQLite URIs so paths with spaces / `#` / `?` open cleanly
3. `workspace_listing.py` + `workspace_tabs.py` — accept dict-shaped `projectLayouts` entries (previously string-only; resolver already handled both)
4. `workspace_resolver.py:_get_project_from_file_path` — use `os.path.commonpath` instead of `startswith` so `/repo/app` doesn't sibling-match `/repo/app2`
5. `workspace_tabs.py` — mark synthetic "Tool Action" bubbles with `synthetic=True` so they don't pollute `responseTimeMs` / `totalResponseTimeMs`

## Test plan

- [x] `python3 -m unittest discover tests -v` — 186 pass
- [x] App boots, all routes register
- [x] Manual smoke against real Cursor data: workspaces list, tabs, search all return expected shapes
- [ ] CI matrix (ubuntu-latest × Python 3.11/3.12/3.13) — pending

## Summary by CodeRabbit

* **Refactor**
  * Moved workspace/project resolution and tab assembly into dedicated service layers for clearer separation and maintainability.
* **New Features**
  * CLI-derived chats now appear as projects and tabs.
  * Improved workspace descriptor handling for varied path/URI formats and indirections.
* **Bug Fixes**
  * Endpoints tolerate malformed session/conversation records, corrupted or missing DBs, and workspace paths with spaces without crashing.
* **Tests**
  * Added extensive unit tests covering CLI sessions, path-boundary cases, DB edge cases, and malformed conversation handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/cppalliance/cppa-cursor-browser/pull/31)

### CODE REVIEW

File: services/workspace_tabs.py [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+                        for layout in layouts:
+                            if isinstance(layout, str):
+                                try:
+                                    layout = json.loads(layout)
+                                except Exception:
+                                    continue
+                            if isinstance(layout, dict) and layout.get("rootPath"):
+                                project_layouts_map[cid].append(layout["rootPath"])
+                except Exception:
+                    pass
```
Comment (bradjin8) [inline services/workspace_tabs.py]: Bug — unguarded sqlite3.Error on all DB queries

File: api/workspaces.py
Comment (bradjin8) [inline api/workspaces.py]: Bug — unsafe dict access in CLI branch of get_workspace

File: services/workspace_listing.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                "id": f"cli:{cp['project_id']}",
+                "name": ws_name,
+                "conversationCount": len(cli_convos),
+                "lastModified": (
+                    datetime.fromtimestamp(last_ms / 1000, tz=timezone.utc).isoformat()
+                    if last_ms
+                    else datetime.now(tz=timezone.utc).isoformat()
+                ),
+                "source": "cli",
+            })
```
Comment (bradjin8) [inline services/workspace_listing.py:275]: Bug — unsafe dict access in CLI project loop

File: services/cli_tabs.py
Comment (bradjin8) [inline services/cli_tabs.py:1]: Line 17, 22, 25:

Bug — unsafe dict access in project lookup

File: services/workspace_tabs.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+
+                # Code block diffs as extra bubbles
+                diffs = code_block_diff_map.get(composer_id, [])
+                for diff in diffs:
+                    diff_text = format_tool_action(diff)
+                    if diff_text.strip():
+                        synthetic_ts = (
+                            to_epoch_ms(diff.get("timestamp"))
+                            or to_epoch_ms(diff.get("createdAt"))
+                            or max((b.get("timestamp") or 0) for b in bubbles)
```
Comment (bradjin8) [inline services/workspace_tabs.py]: Design — diffs double-represented in response

File: services/workspace_tabs.py [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+            if len(parts) >= 2:
+                cid = parts[1]
+                try:
+                    ctx = json.loads(row["value"])
+                    layouts = ctx.get("projectLayouts")
+                    if isinstance(layouts, list):
+                        project_layouts_map.setdefault(cid, [])
+                        for layout in layouts:
+                            if isinstance(layout, str):
+                                try:
```
Comment (bradjin8) [inline services/workspace_tabs.py]: Performance — messageRequestContext queried twice

File: services/workspace_db.py
Comment (bradjin8) [inline services/workspace_db.py:1]: Line 23–24, 37–38, 68–69:

Quality — broad silent except Exception: pass

**Review by bradjin8 — DISMISSED** [review]

File: services/workspace_tabs.py (hunk 1 above)
Comment (timon0305) [reply services/workspace_tabs.py]: it is fixed

Comment (timon0305) [reply services/workspace_tabs.py]: Resolved

File: api/workspaces.py
Comment (timon0305) [reply api/workspaces.py]: fixed

File: services/workspace_listing.py (hunk 2 above)
Comment (timon0305) [reply services/workspace_listing.py:275]: five queued fixed

File: services/cli_tabs.py
Comment (timon0305) [reply services/cli_tabs.py:1]: fixed

File: services/workspace_tabs.py (hunk 3 above)
Comment (timon0305) [reply services/workspace_tabs.py]: okay

File: services/workspace_tabs.py (hunk 4 above)
Comment (timon0305) [reply services/workspace_tabs.py]: Fixed

**Review by bradjin8 — APPROVED** [review]

**Review by wpak-ai — APPROVED** [review]
